gear_menu: Add theme switcher to the gear menu popover.

sayamsamal opened the pull request:
This PR improves the theme experience for spectators by:

* **Adding a Theme Switcher:** Replaces separate light/dark theme options with the new 3-way theme switcher in the gear menu popover, standardizing theme selection across the app.
* **Updating the Realm Logo:** Ensures the realm logo dynamically updates on theme change without requiring a page reload.
* **Supporting Automatic Color Scheme:** Implements automatic color scheme support by storing the user's preference in local storage, mirroring the light/dark theme functionality.

Fixes: #30318

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**

| Desc | Before | After |
|--------|--------|--------|
| Light Mode | ![gear-menu-before-light](https://github.com/zulip/zulip/assets/82862779/263b5810-500f-4807-a299-dac74396c021) | ![gear-menu-after-light](https://github.com/zulip/zulip/assets/82862779/ef8be18f-2a72-430d-ac44-2fe68219cb97) |
| Dark Mode | ![gear-menu-before-dark](https://github.com/zulip/zulip/assets/82862779/749261b2-f9f2-42e0-8827-7795b68a2c69) | ![gear-menu-after-dark](https://github.com/zulip/zulip/assets/82862779/3ea9b89c-6d95-4fa7-a381-19996f9b6888) |
| Dynamic Realm Logo on theme change | ![realm-icon-before](https://github.com/zulip/zulip/assets/82862779/ef68cd8f-515e-4e5a-97cc-0caac5809c65) | ![realm-icon-after](https://github.com/zulip/zulip/assets/82862779/5823a399-c6bc-45cb-812d-1d953d2c68f9) |

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

sayamsamal commented:
@amanagr Could you please review this PR? Thanks!

amanagr commented on web/src/dark_theme.ts:
```diff
@@ -22,9 +24,17 @@ export function disable(): void {
     }
 }
 
 export function default_preference_checker(): void {
```
Should we rename this function and library? `dark_theme.default_preference_checker` is a weird thing to call.

I would prefer it to be called `theme.use_system_default` instead.

amanagr replied:
Let's open a czo thread to weight in opinion from others.

sayamsamal replied:
Opened a [CZO thread](https://chat.zulip.org/#narrow/stream/6-frontend/topic/refactor.20dark_theme.2Ets.20module/near/1821507) discussing this idea, thanks!

amanagr commented on web/src/dark_theme.ts:
```diff
@@ -22,9 +24,17 @@ export function disable(): void {
     $(":root").removeClass("dark-theme").addClass("color-scheme-automatic");
+
+    if (page_params.is_spectator) {
+        const ls = localstorage();
+        ls.set("spectator-theme-preference", "automatic");
+        user_settings.color_scheme = settings_config.color_scheme_values.automatic.code;
+        realm_logo.render();
```
Let's extract this into a function since it seems like it would make this easier to look at and of course has the benefit of DRY.

sayamsamal replied:
Extracted this and other such methods for setting spectator theme into `dark_theme.set_theme_for_spectator()`.

amanagr commented on web/src/gear_menu.js:
```diff
@@ -161,22 +162,16 @@ export function initialize() {
-                e.stopPropagation();
+            $popper.on("change", "input[name='theme-select']", (e) => {
+                const theme_code = Number.parseInt($(e.currentTarget).attr("data-theme-code"), 10);
                 requestAnimationFrame(() => {
-                    dark_theme.disable();
+                    if (theme_code === settings_config.color_scheme_values.night.code) {
+                        dark_theme.enable();
+                    } else if (theme_code === settings_config.color_scheme_values.day.code) {
+                        dark_theme.disable();
+                    } else {
+                        dark_theme.default_preference_checker();
+                    }
                     message_lists.update_recipient_bar_background_color();
```
Let's move this call inside `dark_theme` and calling it in the server events dispatch path too.

sayamsamal replied:
Moved the CSS class logic into `dark_theme.set_theme()` and added `dark_theme.set_theme_and_update()`, which calls `set_theme()` and also calls the realm logo and recipient bar update methods. The idea is that all theme switchers and commands call `dark_theme.set_theme_and_update()`, which will set the theme and also update the necessary elements. While `dark_theme.set_theme()` is used in `ui_init.js` to only set the theme via the CSS classes.

amanagr commented on web/src/dark_theme.ts:
```diff
@@ -12,6 +13,7 @@ export function enable(): void {
         const ls = localstorage();
         ls.set("spectator-theme-preference", "dark");
         user_settings.color_scheme = settings_config.color_scheme_values.night.code;
+        realm_logo.render();
```
I was wondering why we don't need `realm_logo.render();` call for logged-in users and it seems like we do, so we move the call inside these `enable/ disable` functions as well.

sayamsamal replied:
Makes sense, thanks! Moved `realm_logo.render()` along with `message_lists.update_recipient_bar_background_color()` into `dark_theme` module.

sayamsamal commented:
@amanagr this is ready for another review.

Updates:
- Refactored and consolidated theme setting logic.
- Renamed `dark_theme.ts` -> `theme.ts`.
- Renamed `enable` -> `set_dark_theme`, `disable` -> `set_light_theme`, `default_preference_checker` -> `set_automatic_theme.`
- Renamed day->light and night->dark in the frontend code.

amanagr commented:
Looks good to me. Checked that all the feedback is addressed and manually tested for spectators and logged in user for all 3 themes. Checked automatic theme by varying theme in system settings on mac.

timabbott commented:
This looks great, merged, thanks for all the cleanups @sayamsamal!
